### ISSUE TITLE
Recurly::Transaction::DeclinedError#to_s raises an exception

### ISSUE THREAD

**Issue by sdhull** [recurly/recurly-client-ruby#269]
I'm not sure what exactly is going on with `Recurly::XML` and `Recurly::XML::NokogiriAdapter`, but it seems to be breaking.

```
irb(main):058:0> purchase.recurly_error.to_s
TypeError: wrong argument type Nokogiri::XML::Element (expected Data)
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/nokogiri-1.5.11/lib/nokogiri/xml/node.rb:146:in `document'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/nokogiri-1.5.11/lib/nokogiri/xml/node.rb:146:in `xpath'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/nokogiri-1.5.11/lib/nokogiri/xml/node.rb:239:in `at_xpath'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/recurly-2.7.0/lib/recurly/xml/nokogiri.rb:35:in `text'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/recurly-2.7.0/lib/recurly/transaction/errors.rb:39:in `to_s'
        from (irb):58
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/railties-3.2.22.2/lib/rails/commands/console.rb:47:in `start'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/railties-3.2.22.2/lib/rails/commands/console.rb:8:in `start'
        from /var/www/turtle/shared/bundle/ruby/2.2.0/gems/railties-3.2.22.2/lib/rails/commands.rb:41:in `<top (required)>'
        from script/rails:6:in `require'
        from script/rails:6:in `<main>'
irb(main):059:0> purchase.recurly_error.class
=> Recurly::Transaction::DeclinedError
```

This is problematic when trying to do `recurly_error.to_s` as a uniform way to access an appropriate customer-facing error message, regardless of whether they're creating a subscription, a transaction, redeeming a coupon, or whatever.

**Comment by bhelx** [comment]
@sdhull It would be great if I could see the xml being returned here. You can enable debug logging of the xml like this:

```ruby
require 'logger'
Recurly.logger = Logger.new STDOUT
```

Also, if I could get some kind of context code, or a script to reproduce, that would be helpful.

At first glance, it looks like maybe this path doesn't exist https://github.com/recurly/recurly-client-ruby/blob/master/lib/recurly/transaction/errors.rb#L39 but i'm not sure yet.

**Comment by sdhull** [comment]
So we're serializing & storing exceptions from recurly so we can inspect them later (hence `purchase.recurly_error`).

Just reproduced it easily by going through the flow locally and using the [test gateway](https://docs.recurly.com/docs/test) with one of the declined cards. Here's the XML:

```
<?xml version="1.0" encoding="UTF-8"?>
<errors>
  <transaction_error>
    <error_code>declined</error_code>
    <error_category>soft</error_category>
    <merchant_message>The customer's bank has declined their card. The customer will need to contact their bank to learn the cause.</merchant_message>
    <customer_message>Your transaction was declined. Please use a different card or contact your bank.</customer_message>
    <gateway_error_code nil="nil"></gateway_error_code>
  </transaction_error>
  <error field="billing_info.base" symbol="declined">Your transaction was declined. Please use a different card or contact your bank.</error>
  <transaction href="https://looksharp-staging.recurly.com/v2/transactions/3951e28e9474667ba43a8644808320f9" type="credit_card">
    <account href="https://looksharp-staging.recurly.com/v2/accounts/org21639"/>
    <uuid>3951e28e9474667ba43a8644808320f9</uuid>
    <action>verify</action>
    <amount_in_cents type="integer">0</amount_in_cents>
    <tax_in_cents type="integer">0</tax_in_cents>
    <currency>USD</currency>
    <status>declined</status>
    <payment_method>credit_card</payment_method>
    <reference>8738574</reference>
    <source>billing_info</source>
    <recurring type="boolean">false</recurring>
    <test type="boolean">true</test>
    <voidable type="boolean">false</voidable>
    <refundable type="boolean">false</refundable>
    <ip_address>76.216.1.36</ip_address>
    <transaction_error>
      <error_code>declined</error_code>
      <error_category>soft</error_category>
      <merchant_message>The customer's bank has declined their card. The customer will need to contact their bank to learn the cause.</merchant_message>
      <customer_message>Your transaction was declined. Please use a different card or contact your bank.</customer_message>
      <gateway_error_code nil="nil"></gateway_error_code>
    </transaction_error>
    <cvv_result code="" nil="nil"></cvv_result>
    <avs_result code="" nil="nil"></avs_result>
    <avs_result_street nil="nil"></avs_result_street>
    <avs_result_postal nil="nil"></avs_result_postal>
    <created_at type="datetime">2016-10-20T17:15:38Z</created_at>
    <updated_at type="datetime">2016-10-20T17:15:38Z</updated_at>
    <details>
      <account>
        <account_code>org21639</account_code>
        <first_name>Foo</first_name>
        <last_name>Bar</last_name>
        <company>Foobar</company>
        <email>foo@davidverba.com</email>
        <billing_info type="credit_card">
          <first_name>Joe</first_name>
          <last_name>SHmoe</last_name>
          <address1>123 Main St</address1>
          <address2 nil="nil"></address2>
          <city nil="nil"></city>
          <state nil="nil"></state>
          <zip>94605</zip>
          <country>US</country>
          <phone nil="nil"></phone>
          <vat_number nil="nil"></vat_number>
          <card_type>Visa</card_type>
          <year type="integer">2021</year>
          <month type="integer">12</month>
          <first_six>400000</first_six>
          <last_four>0002</last_four>
        </billing_info>
      </account>
    </details>
  </transaction>
</errors>
```

IMO, it would be easier to remove the monkey-patching of nokogiri and just use the library as-is. Should be sufficient.

**Comment by sdhull** [comment]
To be clear, this isn't a "once in awhile" thing, this is an "every time" thing. Every time you get a transaction declined exception, if you try to do `exception.to_s` it blows up.

**Comment by bhelx** [comment]
Okay, sorry about that. I'm not sure I can reproduce it though. Here is the code i tried to use to trigger it:

```ruby
require 'securerandom'

begin
  account_code = SecureRandom.uuid
  plan_code = Recurly::Plan.all.first.plan_code

  Recurly::Subscription.create(
    :plan_code => plan_code,
    :account   => {
      :account_code => account_code,
      :billing_info => {
        :number => '4000-0000-0000-0002', # failing card
        :month  => 1,
        :year   => 2019,
        :state => "LA",
        :country => "US",
        :address1 => "400 alabama st"
      }
    }
  )
rescue => e
  puts e.inspect
  puts e.to_s
end
```

Results in:

```
#<Recurly::Transaction::DeclinedError: Your transaction was declined. Please use a different card or contact your bank.>
Your transaction was declined. Please use a different card or contact your bank.
```

**Comment by bhelx** [comment]
Could it possibly be a nokogiri versioning issue?

**Comment by sdhull** [comment]
Strange. I suppose it could be a version issue. `nokogiri (1.5.11)`

One difference is we're using Recurly.js to get a token to update billing info and then using the account_code for the transaction, but I don't expect that to really be much of an issue.

What version of Nokogiri are you using? Does the XML I posted parse properly for you?

**Comment by bhelx** [comment]
Well, I can't say for sure if that's the problem, but you'll want to try upgrading nokogiri to at least 1.6. I'm testing on `1.6.6`. There is a fixture we are using in the tests and it looks the same as your xml as far as I can tell: https://github.com/recurly/recurly-client-ruby/blob/master/spec/fixtures/transaction_error.xml

My live response looks like this:

```xml
<errors>
  <transaction_error>
    <error_code>declined</error_code>
    <error_category>soft</error_category>
    <merchant_message>The customer's bank has declined their card. The customer will need to contact their bank to learn the cause.</merchant_message>
    <customer_message>Your transaction was declined. Please use a different card or contact your bank.</customer_message>
    <gateway_error_code nil="nil"></gateway_error_code>
  </transaction_error>
  <error field="subscription.account.base" symbol="declined">Your transaction was declined. Please use a different card or contact your bank.</error>
  <transaction href="https://bhelx.recurly.com/v2/transactions/3952793f3723422a3349d64888b5dff1" type="credit_card">
    <subscription href="https://bhelx.recurly.com/v2/subscriptions/3952793ec0f7712846fac04d78bd26dc"/>
    <uuid>3952793f3723422a3349d64888b5dff1</uuid>
    <action>purchase</action>
    <amount_in_cents type="integer">2000</amount_in_cents>
    <tax_in_cents type="integer">0</tax_in_cents>
    <currency>USD</currency>
    <status>declined</status>
    <payment_method>credit_card</payment_method>
    <reference>8548026</reference>
    <source>subscription</source>
    <recurring type="boolean">false</recurring>
    <test type="boolean">true</test>
    <voidable type="boolean">false</voidable>
    <refundable type="boolean">false</refundable>
    <ip_address nil="nil"></ip_address>
    <transaction_error>
      <error_code>declined</error_code>
      <error_category>soft</error_category>
      <merchant_message>The customer's bank has declined their card. The customer will need to contact their bank to learn the cause.</merchant_message>
      <customer_message>Your transaction was declined. Please use a different card or contact your bank.</customer_message>
      <gateway_error_code nil="nil"></gateway_error_code>
    </transaction_error>
    <cvv_result code="" nil="nil"></cvv_result>
    <avs_result code="" nil="nil"></avs_result>
    <avs_result_street nil="nil"></avs_result_street>
    <avs_result_postal nil="nil"></avs_result_postal>
    <created_at type="datetime">2016-10-20T20:00:13Z</created_at>
    <updated_at type="datetime">2016-10-20T20:00:13Z</updated_at>
    <details>
      <account>
        <account_code>3e9da6e0-5791-4522-b849-4d085ab52d07</account_code>
        <first_name>Verena</first_name>
        <last_name>Example</last_name>
        <company nil="nil"></company>
        <email>verena@example.com</email>
        <billing_info type="credit_card">
          <first_name>Verena</first_name>
          <last_name>Example</last_name>
          <address1>400 alabama st</address1>
          <address2 nil="nil"></address2>
          <city nil="nil"></city>
          <state>LA</state>
          <zip nil="nil"></zip>
          <country>US</country>
          <phone nil="nil"></phone>
          <vat_number nil="nil"></vat_number>
          <card_type>Visa</card_type>
          <year type="integer">2019</year>
          <month type="integer">1</month>
          <first_six>400000</first_six>
          <last_four>0002</last_four>
        </billing_info>
      </account>
    </details>
  </transaction>
</errors>
```

I didn't try parsing your xml directly but i can give it a shot.

**Comment by sdhull** [comment]
Unfortunately, ours is a legacy project and upgrading Nokogiri is not a simple task, thanks to dependency hell. 😞

**Comment by sdhull** [comment]
```
    fog was resolved to 1.38.0, which depends on
      fog-vsphere (>= 0.4.0) was resolved to 1.4.0, which depends on
        rbvmomi (~> 1.9) was resolved to 1.9.4, which depends on
          nokogiri (~> 1.5)
```
Which is, amazingly, the latest version of fog. How can that be?

**Comment by sdhull** [comment]
OK took some doing but I got Nokogiri updated to `1.6.8.1` and it resolved the issue.
